Bug Report: mysqld socket file length breaks unit test Mac OS

### Overview of the Issue

Hello, first time poster here :) 

I'm trying to get `make unit_test` to run on Mac OS. Got a failure with `./go/cmd/vttestserver/vttestserver_test.go`, specifically with `TestPersistentMode`.

I believe the issue is that `t.TempDir()` on Mac OS generates a pretty long directory name (e.g. `/var/folders/96/k7gzd7q10zdb749vr02q7sjh0000gn/T/TestPersistentMode2756112461`), and when MySQL tries to connect to a socket stored in a path that long, it fails.

I was able to work around this by forcing the test to use a shorter file name.

```
diff --git a/go/cmd/vttestserver/vttestserver_test.go b/go/cmd/vttestserver/vttestserver_test.go
index 5b9139987e..1c2a7c3182 100644
--- a/go/cmd/vttestserver/vttestserver_test.go
+++ b/go/cmd/vttestserver/vttestserver_test.go
@@ -75,7 +75,7 @@ func TestPersistentMode(t *testing.T) {
        conf := config
        defer resetFlags(args, conf)
 
-       dir := t.TempDir()
+       dir := "/tmp/vitess_TestPersistentMode" //t.TempDir()
 
        cluster, err := startPersistentCluster(dir)
        assert.NoError(t, err)
```

### Reproduction Steps



Run all unit tests:
```
make unit_test
```

### Binary Version

 * [Vitess@3f79a0ea5](https://github.com/vitessio/vitess/commit/3f79a0ea50374aa0647c7497c821850afd538040)
 * MySQL 8.0.29, installed with `brew install mysql@8.0`

### Operating System and Environment details

```sh
➜  vitess git:(main) ✗ cat /etc/os-release
cat: /etc/os-release: No such file or directory
➜  vitess git:(main) ✗ uname -sr
Darwin 21.3.0
➜  vitess git:(main) ✗ uname -m
arm64
```

Additionally:
 * MacBook Air (M1 ,2020), 16 GB
 * macOS Monterrey 12.2.1

### Log Fragments

```sh
I0707 11:29:55.281313   95099 mysqld.go:386] Mysqld.Start(1657207795) &exec.Cmd{Path:"/Users/maxenglander/dev/homebrew/homebrew/bin/mysqld_safe", Args:[]string{"/Users/maxenglander/dev/homebrew/homebrew/bin/mysqld_safe", "--defaults-file=/var/folders/96/k7gzd7q10zdb749vr02q7sjh0000gn/T/TestPersistentMode1174413336/001/vt_0000000001/my.cnf", "--basedir=/Users/maxenglander/dev/homebrew/homebrew"}, Env:[]string{"LD_LIBRARY_PATH=/Users/maxenglander/dev/homebrew/homebrew/lib/mysql", "LD_PRELOAD="}, Dir:"/Users/maxenglander/dev/homebrew/homebrew", Stdin:io.Reader(nil), Stdout:io.Writer(nil), Stderr:io.Writer(nil), ExtraFiles:[]*os.File(nil), SysProcAttr:(*syscall.SysProcAttr)(nil), Process:(*os.Process)(nil), ProcessState:(*os.ProcessState)(nil), ctx:context.Context(nil), lookPathErr:error(nil), finished:false, childFiles:[]*os.File(nil), closeAfterStart:[]io.Closer(nil), closeAfterWait:[]io.Closer(nil), goroutine:[]func() error(nil), errch:(chan error)(nil), waitDone:(chan struct {})(nil)}
I0707 11:29:55.291720   95099 mysqld.go:453] Waiting for mysqld socket file (/var/folders/96/k7gzd7q10zdb749vr02q7sjh0000gn/T/TestPersistentMode1174413336/001/vt_0000000001/mysql.sock) to be ready...
I0707 11:30:00.341948   95099 mysqld.go:404] Mysqld.Start(1657207795) stdout: 2022-07-07T15:30:00.6NZ mysqld_safe Logging to '/var/folders/96/k7gzd7q10zdb749vr02q7sjh0000gn/T/TestPersistentMode1174413336/001/vt_0000000001/error.log'.
I0707 11:30:00.690677   95099 mysqld.go:404] Mysqld.Start(1657207795) stdout: 2022-07-07T15:30:00.6NZ mysqld_safe Starting mysqld daemon with databases from /var/folders/96/k7gzd7q10zdb749vr02q7sjh0000gn/T/TestPersistentMode1174413336/001/vt_0000000001/data
I0707 11:30:02.524253   95099 mysqld.go:404] Mysqld.Start(1657207795) stdout: 2022-07-07T15:30:02.6NZ mysqld_safe mysqld from pid file /var/folders/96/k7gzd7q10zdb749vr02q7sjh0000gn/T/TestPersistentMode1174413336/001/vt_0000000001/mysql.pid ended
I0707 11:30:02.540622   95099 mysqld.go:417] Mysqld.Start(1657207795) exit: <nil>
--- FAIL: TestPersistentMode (60.03s)
    vttestserver_test.go:81:
                Error Trace:    vttestserver_test.go:81
                Error:          Received unexpected error:
                                signal: killed
                Test:           TestPersistentMode
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x1cf35d6]

goroutine 216 [running]:
testing.tRunner.func1.2({0x1dfefc0, 0x3085910})
        /usr/local/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1392 +0x39f
panic({0x1dfefc0, 0x3085910})
        /usr/local/go/src/runtime/panic.go:838 +0x207
vitess.io/vitess/go/vt/vttest.(*LocalCluster).GrpcPort(...)
        /Users/maxenglander/dev/vitessio/vitess/go/vt/vttest/local_cluster.go:570
vitess.io/vitess/go/cmd/vttestserver.assertColumnVindex(_, {{0xc000112c60, 0x0, {0x1fca75e, 0xb}, {0x0, 0x0}, {0xc00049212b, 0x51}, {0x1fb47d8, ...}, ...}, ...}, ...)
        /Users/maxenglander/dev/vitessio/vitess/go/cmd/vttestserver/vttestserver_test.go:341 +0x36
vitess.io/vitess/go/cmd/vttestserver.TestPersistentMode(0xc0001316c0)
        /Users/maxenglander/dev/vitessio/vitess/go/cmd/vttestserver/vttestserver_test.go:84 +0x265
testing.tRunner(0xc0001316c0, 0x22aaaf8)
        /usr/local/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1486 +0x35f
FAIL    vitess.io/vitess/go/cmd/vttestserver    77.086s
```

--- a/tools/unit_test_runner.sh
+++ b/tools/unit_test_runner.sh
@@ -36,6 +36,20 @@ if [[ -z $VT_GO_PARALLEL && -n $VT_GO_PARALLEL_VALUE ]]; then
   VT_GO_PARALLEL="-p $VT_GO_PARALLEL_VALUE"
 fi
 
+# Mac makes long temp directories for os.TempDir(). MySQL can't connect to
+# sockets in those directories. Tell Golang to use /tmp/vttest_XXXXXX instead.
+kernel="$(uname -s)"
+case "$kernel" in
+  darwin|Darwin)
+    TMPDIR=${TMPDIR:-}
+    if [ -z "$TMPDIR" ]; then
+      TMPDIR="$(mktemp -d /tmp/vttest_XXXXXX)"
+      export TMPDIR
+    fi
+    echo "Using temporary directory for tests: $TMPDIR"
+    ;;
+esac
+
 # All Go packages with test files.
 # Output per line: <full Go package name> <all _test.go files in the package>*
 packages_with_tests=$(go list -f '{{if len .TestGoFiles}}{{.ImportPath}} {{join .TestGoFiles " "}}{{end}}' ./go/... | sort)

